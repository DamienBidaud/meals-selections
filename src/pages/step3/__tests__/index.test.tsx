import { render, RenderResult, screen, waitFor } from "@testing-library/react";
import { useEffect } from "react";
import { atom, RecoilRoot, useRecoilValue } from "recoil";
import Step3 from "../index";
import { createBrowserRouter, RouterProvider } from "react-router-dom";
import * as routerFunction from "react-router";
import userEvent from "@testing-library/user-event";

//@ts-ignore
export const RecoilObserver = ({ node, onChange }) => {
  const value = useRecoilValue(node);
  useEffect(() => onChange(value), [onChange, value]);
  return null;
};

const mockNavigate = jest.fn();
const mockUseNavigate = jest.spyOn(routerFunction, "useNavigate");

const router = createBrowserRouter([
  {
    path: "/",
    element: <Step3 />,
  },
]);
const formState = atom({
  key: "form",
  default: {
    guests: 3,
    meal: "dinner",
    restaurant: "Mc Donalds",
    selectedMeals: [
      {
        meal: "",
        serving: 1,
      },
    ],
  },
});

const onChange = jest.fn();
describe("Step 3", () => {
  let step3: RenderResult;
  beforeEach(() => {
    step3 = render(
      <RecoilRoot>
        <RouterProvider router={router} />
        <RecoilObserver node={formState} onChange={onChange} />
      </RecoilRoot>
    );
  });

  afterEach(() => {
    jest.resetAllMocks();
  });
  it("Should match the snapshot", () => {
    const { asFragment } = step3;
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should not navagitate when no restaurant have been selected", () => {
    const alertMock = jest.spyOn(window, "alert").mockImplementation();
    mockUseNavigate.mockReturnValue(mockNavigate);

    userEvent.click(screen.getByText("Next"));
    expect(mockNavigate).not.toHaveBeenCalled();
    expect(alertMock).toHaveBeenCalledTimes(1);
  });

  it("Should navigate when a restaurant have been selected", () => {
    const alertMock = jest.spyOn(window, "alert").mockImplementation();
    mockUseNavigate.mockReturnValue(mockNavigate);
    userEvent.selectOptions(screen.getByRole("combobox"), "Fries");

    userEvent.click(screen.getByText("Next"));
    expect(onChange).toHaveBeenCalledWith({
      guests: 3,
      meal: "dinner",
      restaurant: "Mc Donalds",
      selectedMeals: [
        {
          meal: "Fries",
          serving: 3,
        },
      ],
    });
    expect(mockNavigate).toHaveBeenCalledWith("/step4");
    expect(alertMock).toHaveBeenCalledTimes(0);
  });
});
