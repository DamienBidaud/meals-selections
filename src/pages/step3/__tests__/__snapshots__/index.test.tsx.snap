// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step 3 Should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe bMEgX"
  >
    <h1
      class="sc-eDvSVe koGjPR"
    >
      Step 3
    </h1>
    <div>
      <div
        class="sc-dkrFOg iKbvez"
      >
        <div
          class="sc-hLBbgP kHbnCM"
        >
          <label
            for="dishList"
          >
            Please select a dish
          </label>
          <select
            id="dishList"
            name="dish"
          >
            <option
              disabled=""
              selected=""
              value=""
            >
              ----
            </option>
            <option
              value="Chicken Burger"
            >
              Chicken Burger
            </option>
            <option
              value="Ham Burger"
            >
              Ham Burger
            </option>
            <option
              value="Cheese Burger"
            >
              Cheese Burger
            </option>
            <option
              value="Fries"
            >
              Fries
            </option>
          </select>
        </div>
        <div>
          <label
            for="serving"
          >
            Please enter number of serving
          </label>
          <input
            id="serving"
            max="10"
            min="3"
            type="number"
            value="3"
          />
        </div>
        <hr />
      </div>
    </div>
    <div
      class="sc-gswNZR hNVRgE"
    >
      <button
        class="sc-jSUZER bqFJJK"
      >
        +
      </button>
      <button
        class="sc-jSUZER bqFJJK"
        disabled=""
      >
        -
      </button>
    </div>
    <div>
      <button
        class="sc-jSUZER bqFJJK"
      >
        Back
      </button>
      <button
        class="sc-jSUZER bqFJJK"
      >
        Next
      </button>
    </div>
  </div>
</DocumentFragment>
`;
